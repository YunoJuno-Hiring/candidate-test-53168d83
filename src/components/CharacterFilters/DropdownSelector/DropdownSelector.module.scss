.DropdownSelector {
  display: flex;
  width: 150px;
  height: 30px;
  margin: 8px 0px;
}

.DropdownSelector-label {
  display: flex;
  font-size: small;
  margin: 0;
  align-self: start;
}
