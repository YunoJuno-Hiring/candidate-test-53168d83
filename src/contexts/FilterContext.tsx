import { IFilterContext } from '@custom-types/contexts';
import React, { ReactNode, createContext, useContext } from 'react';

interface FilterContextProps {
  selectedValues: IFilterContext;
  setSelectedValues: React.Dispatch<React.SetStateAction<IFilterContext>>;
  children?: ReactNode;
}

const FilterContext = createContext<FilterContextProps | undefined>(undefined);

export default function FilterContextProvider({ children }: { children: ReactNode }) {
  const [selectedValues, setSelectedValues] = React.useState<IFilterContext>({
    category: 'all',
    orderBy: 'alphabetical'
  });

  return (
    <FilterContext.Provider value={{ selectedValues, setSelectedValues }}>
      {children}
    </FilterContext.Provider>
  );
}

export const useFilterContext = () => {
  const context = useContext(FilterContext);
  if (!context) {
    throw new Error('useFilterContext must be used within a FilterContextProvider');
  }
  return context;
};
