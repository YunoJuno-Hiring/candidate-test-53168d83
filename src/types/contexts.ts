import { orderByStrings } from '@constants/strings';
import { TCharacterCategories } from './character';

export type TOrderBy = (typeof orderByStrings)[keyof typeof orderByStrings];

export type TContextValue = 'category' | 'orderBy';

export interface IFilterContext {
  category: TCharacterCategories | 'all';
  orderBy: TOrderBy;
}
