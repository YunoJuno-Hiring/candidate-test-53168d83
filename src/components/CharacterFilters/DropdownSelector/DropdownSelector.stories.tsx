import React from 'react';

import type { Meta, StoryObj } from '@storybook/react';
import DropdownSelector from './DropdownSelector';
import { TContextValue } from '@custom-types/contexts';
import FilterContextProvider from '@contexts/FilterContext';

const meta: Meta<typeof DropdownSelector> = {
  component: DropdownSelector,
  decorators: [
    (Story) => (
      <FilterContextProvider>
        <Story />
      </FilterContextProvider>
    )
  ],
  tags: ['autodocs']
};

export default meta;

type Story = StoryObj<typeof DropdownSelector>;

export const Default: Story = {
  args: {
    selectOptions: ['one', 'two', 'three'],
    contextValue: 'category' as TContextValue,
    label: 'Order by'
  }
};
