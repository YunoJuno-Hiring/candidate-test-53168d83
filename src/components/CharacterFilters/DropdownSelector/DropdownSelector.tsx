// DropdownSelector.tsx
import { useFilterContext } from '@contexts/FilterContext';
import React from 'react';
import { capitaliseFirstCharacters } from '@utils/stringHelpers';
import { TContextValue } from '@custom-types/contexts';
import styles from './DropdownSelector.module.scss';

interface DropdownSelectorProps {
  selectOptions: string[];
  contextValue: TContextValue;
  label?: string;
}
export default function DropdownSelector({
  selectOptions,
  contextValue,
  label
}: DropdownSelectorProps) {
  const { selectedValues, setSelectedValues } = useFilterContext();

  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setSelectedValues({ ...selectedValues, [contextValue]: event.target.value });
  };

  return (
    <div className={styles['DropdownSelector-container']}>
      <p className={styles['DropdownSelector-label']}>{label}</p>
      <select
        className={styles['DropdownSelector']}
        value={selectedValues[contextValue]}
        onChange={handleChange}
      >
        {selectOptions.map((option: string, key: number) => (
          <option value={option} key={key}>
            {capitaliseFirstCharacters(option)}
          </option>
        ))}
      </select>
    </div>
  );
}
