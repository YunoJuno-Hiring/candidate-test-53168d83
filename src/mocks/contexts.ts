import { IFilterContext } from '@custom-types/contexts';

export function generateMockFilterContext(overrides?: Partial<IFilterContext>): IFilterContext {
  const defaultContext = {
    category: 'all',
    orderBy: 'significance'
  };

  return { ...defaultContext, ...overrides } as IFilterContext;
}
