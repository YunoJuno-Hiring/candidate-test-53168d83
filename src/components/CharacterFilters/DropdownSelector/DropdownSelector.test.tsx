import FilterContextProvider from '@contexts/FilterContext';
import { TContextValue } from '@custom-types/contexts';
import React from 'react';
import renderer from 'react-test-renderer';
import DropdownSelector from './DropdownSelector';

const mockDropdownProps = {
  selectOptions: ['One', 'Two', 'Three'],
  contextValue: 'category' as TContextValue
};

describe('DropdownSelector', () => {
  it('Should match the DOM snapshot', () => {
    const component = renderer
      .create(
        <FilterContextProvider>
          <DropdownSelector {...mockDropdownProps} />
        </FilterContextProvider>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
