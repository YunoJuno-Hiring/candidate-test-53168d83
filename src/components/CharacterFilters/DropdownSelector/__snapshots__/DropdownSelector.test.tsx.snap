// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownSelector Should match the DOM snapshot 1`] = `
<div>
  <p
    className="DropdownSelector-label"
  />
  <select
    className="DropdownSelector"
    onChange={[Function]}
    value="all"
  >
    <option
      value="One"
    >
      One
    </option>
    <option
      value="Two"
    >
      Two
    </option>
    <option
      value="Three"
    >
      Three
    </option>
  </select>
</div>
`;
